UserPasswords: Replace salted SHA-1 with bcrypt

oliverguenther opened the pull request:
This adds a typed `UserPassword` class for Bcrypt password hashes
that replace the previous default of salted SHA1.

It still provides a class for SHA1 passwords, however new passwords of
that type may no longer be created directly, and passwords are checked
whether to upgrade them on every login.

https://community.openproject.com/work_packages/22769

oliverguenther committed:
UserPasswords: Replace salted SHA-1 with bcrypt

This adds a typed userpassword class for Bcrypt password hashes
that replace the previous default of salted SHA1.

It still provides a class for SHA1 passwords, however new passwords of
that type may no longer be created directly, and passwords are checked
whether to upgrade them on every login.

machisuji commented on app/models/user_password.rb:
```diff
@@ -48,39 +79,25 @@ def expired?
-    res = 0
-    b.each_byte do |byte| res |= byte ^ l.shift end
-    res == 0
+  def generate_salt
```
Why is this still here? As far as I can tell this is no longer used anywhere. Also it's not implemented by either subclass.

oliverguenther committed:
Actually rehash on the correct object instance

`becomes!` does not change the current object (how would it?), so we
need to call `hash_password!` on the object returned from it.

This also extends the false positive test. The rehash actually happened,
but it was hashing the plain as a legacy SHA-1 again.

oliverguenther committed:
Remove unused method

machisuji commented:
👍
